### PR TITLE
feat (coupons): prevent applying coupon if there is plan overlap

### PR DESCRIPTION
**Pull request by lovrocolic**
## Context

In some cases coupons needs to be applied only on certain charges. In order to support it, billable metric limitations will be added.

## Description

Prevent applying coupon if there is plan overlap.
There is plan overlap if either plans or BMs directly overlap. Also, there can be overlap if billable metric limitation is present on one coupon but at the same time there is plan limitation on another coupon where plan's charge contain same BM

### CODE REVIEW

File: app/services/applied_coupons/create_service.rb [hunk 1]
```diff
@@ -70,13 +70,41 @@ def reusable_coupon?
...
+      applied_coupons = customer
         .applied_coupons
         .active
         .joins(coupon: :coupon_targets)
+
+      applied_coupons
         .where(coupon_targets: { plan_id: coupon.coupon_targets.select(:plan_id) })
+        .or(
+          applied_coupons
+            .where(coupon_targets: { billable_metric_id: coupon.coupon_targets.select(:billable_metric_id) })
```
Comment (jdenquin) [inline app/services/applied_coupons/create_service.rb:84]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

File: app/services/applied_coupons/create_service.rb [hunk 2]
```diff
@@ -70,13 +70,41 @@ def reusable_coupon?
...
+        )
+        .or(
+          applied_coupons
+            .where(coupon_targets: {
+              plan_id:
+                Charge
+                  .joins(:billable_metric)
+                  .where(billable_metric: { id: coupon.coupon_targets.select(:billable_metric_id) })
+                  .select(:plan_id)
+              }
```
Comment (jdenquin) [inline app/services/applied_coupons/create_service.rb:94]: Layout/FirstHashElementIndentation: Indent the right brace the same as the start of the line where the left brace is.

Comment (jdenquin) [inline app/services/applied_coupons/create_service.rb:94]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

File: app/services/applied_coupons/create_service.rb [hunk 3]
```diff
@@ -70,13 +70,41 @@ def reusable_coupon?
...
+            )
```
Comment (jdenquin) [inline app/services/applied_coupons/create_service.rb:95]: Layout/ClosingParenthesisIndentation: Align `)` with `(`.

Comment (jdenquin) [inline app/services/applied_coupons/create_service.rb:95]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

File: app/services/applied_coupons/create_service.rb [hunk 4]
```diff
@@ -70,13 +70,41 @@ def reusable_coupon?
...
+        )
+        .or(
+          applied_coupons
+            .where(coupon_targets: {
+              billable_metric_id:
+                Charge
+                  .joins(:plan)
+                  .where(plan: { id: coupon.coupon_targets.select(:plan_id) })
+                  .select(:billable_metric_id)
```
Comment (jdenquin) [inline app/services/applied_coupons/create_service.rb:104]: Style/TrailingCommaInHashLiteral: Put a comma after the last item of a multiline hash.

File: app/services/applied_coupons/create_service.rb [hunk 5]
```diff
@@ -70,13 +70,41 @@ def reusable_coupon?
...
+              }
```
Comment (jdenquin) [inline app/services/applied_coupons/create_service.rb:105]: Layout/FirstHashElementIndentation: Indent the right brace the same as the start of the line where the left brace is.

File: app/services/applied_coupons/create_service.rb [hunk 6]
```diff
@@ -70,13 +70,41 @@ def reusable_coupon?
...
+            )
```
Comment (jdenquin) [inline app/services/applied_coupons/create_service.rb:106]: Layout/MultilineMethodCallBraceLayout: Closing method call brace must be on the same line as the last argument when opening brace is on the same line as the first argument.

Comment (jdenquin) [inline app/services/applied_coupons/create_service.rb:106]: Style/TrailingCommaInArguments: Put a comma after the last parameter of a multiline method call. (https://rubystyle.guide#no-trailing-params-comma)

**Review by vincent-pochet — APPROVED** [review]

**Pull request merged**